File: src/test/java/org/dbtools/query/shared/filter/InFilterTest.java
package org.dbtools.query.shared.filter;

import org.dbtools.query.jpa.JPAQueryBuilder;
import org.dbtools.query.shared.QueryBuilder;
import org.dbtools.query.sql.SQLQueryBuilder;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;

import static org.junit.Assert.assertEquals;

public class InFilterTest {

    @Test
    public void testFilter() throws Exception {
        InFilter filter1 = InFilter.create("A", "B");
        assertEquals("Normal Test", "A IN (B)", filter1.buildFilter(new SQLQueryBuilder()));

        InFilter filter2 = InFilter.create("A", Arrays.asList("B", "C", "D"));
        assertEquals("Test List", "A IN (B, C, D)", filter2.buildFilter(new SQLQueryBuilder()));

        InFilter filter3 = InFilter.create("A", (Object)null);
        assertEquals("Test Null", "A IN (null)", filter3.buildFilter(new SQLQueryBuilder()));

        InFilter filter4 = InFilter.create("A", false, "B");
        assertEquals("Normal Test", "A NOT IN (B)", filter4.buildFilter(new SQLQueryBuilder()));

        InFilter filter5 = InFilter.create("A", false, Arrays.asList("B", "C", "D"));
        assertEquals("Test List", "A NOT IN (B, C, D)", filter5.buildFilter(new SQLQueryBuilder()));

        InFilter filter6 = InFilter.create("A", false, (Object)null);
        assertEquals("Test Null", "A NOT IN (null)", filter6.buildFilter(new SQLQueryBuilder()));

        QueryBuilder subQuery = new SQLQueryBuilder().field("Z").table("Y");
        InFilter filter7 = InFilter.create("A", subQuery);
        assertEquals("Test QueryBuilder Implicit", "A IN (SELECT Z FROM Y)", filter7.buildFilter(new SQLQueryBuilder()));

        InFilter filter8 = InFilter.create("A", true, subQuery);
        assertEquals("Test QueryBuilder Explicit", "A IN (SELECT Z FROM Y)", filter8.buildFilter(new SQLQueryBuilder()));

        InFilter filter9 = InFilter.create("A", false, subQuery);
        assertEquals("Test QueryBuilder", "A NOT IN (SELECT Z FROM Y)", filter9.buildFilter(new SQLQueryBuilder()));

    }

    @Test
    public void testAndFilter() throws Exception {
        QueryBuilder subQuery = new SQLQueryBuilder().field("Z").table("Y");
        InFilter filter = InFilter.create("A", false, "B").and("A", true, Arrays.asList("B", "C", "D")).and("A", (Object) null).and("A", subQuery);
        assertEquals("A NOT IN (B) AND A IN (B, C, D) AND A IN (null) AND A IN (SELECT Z FROM Y)", filter.buildFilter(new SQLQueryBuilder()));
    }

    @Test
    public void testOrFilter() throws Exception {
        QueryBuilder subQuery = new SQLQueryBuilder().field("Z").table("Y");
        InFilter filter = InFilter.create("A", "B").or("A", true, Arrays.asList("B", "C", "D")).or("A", false, (Object) null).or("C", "D").or("A", subQuery);
        assertEquals("A IN (B) OR A IN (B, C, D) OR A NOT IN (null) OR C IN (D) OR A IN (SELECT Z FROM Y)", filter.buildFilter(new SQLQueryBuilder()));
    }

    @Test
    public void testComplexFilter() throws Exception {
        QueryBuilder subQuery = new SQLQueryBuilder().field("Z").table("Y");
        InFilter filter = InFilter.create("A", "B").or("A", false, Arrays.asList("B", "C", "D")).or("A", true, subQuery).and("A", (Object) null)
                .and("A", false, subQuery);
        assertEquals("(A IN (B) OR A NOT IN (B, C, D) OR A IN (SELECT Z FROM Y)) AND A IN (null) AND A NOT IN (SELECT Z FROM Y)",
                filter.buildFilter(new SQLQueryBuilder()));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testBadArgument() throws Exception {
        new InFilter("A", true, new ArrayList<Object>());
    }

    @Test
    public void testClone() throws Exception {
        JPAQueryBuilder subQuery = new JPAQueryBuilder();
        subQuery.object("Y", "Y");
        subQuery.field("Y", "Z");
        InFilter filter = InFilter.create("A", "B").and("A", subQuery);
        InFilter clone = filter.clone();
        assertEquals(filter.buildFilter(new JPAQueryBuilder()), clone.buildFilter(new JPAQueryBuilder()));
    }

}

File: dbtools-query/src/commonTest/kotlin/org/dbtools/query/shared/filter/InFilterTest.kt
package org.dbtools.query.shared.filter

import assertk.assertFailure
import org.dbtools.query.sql.SQLQueryBuilder
import kotlin.test.Test
import kotlin.test.assertEquals

class InFilterTest {
    @Test
    fun testFilter() {
        val filter1 = InFilter.create("A", "B")
        assertEquals("A IN (B)", filter1.buildFilter(SQLQueryBuilder()))

        val filter2 = InFilter.create("A", listOf("B", "C", "D"))
        assertEquals("A IN (B, C, D)", filter2.buildFilter(SQLQueryBuilder()))

        val filter3 = InFilter.create("A", null.toString())
        assertEquals("A IN (null)", filter3.buildFilter(SQLQueryBuilder()))

        val filter4 = InFilter.create("A", false, "B")
        assertEquals("A NOT IN (B)", filter4.buildFilter(SQLQueryBuilder()))

        val filter5 = InFilter.create("A", false, listOf("B", "C", "D"))
        assertEquals("A NOT IN (B, C, D)", filter5.buildFilter(SQLQueryBuilder()))

        val filter6 = InFilter.create("A", false, null.toString())
        assertEquals("A NOT IN (null)", filter6.buildFilter(SQLQueryBuilder()))

        val subQuery = SQLQueryBuilder().field("Z").table("Y")
        val filter7 = InFilter.create("A", subQuery)
        assertEquals("A IN (SELECT Z FROM Y)", filter7.buildFilter(SQLQueryBuilder()))

        val filter8 = InFilter.create("A", true, subQuery)
        assertEquals("A IN (SELECT Z FROM Y)", filter8.buildFilter(SQLQueryBuilder()))

        val filter9 = InFilter.create("A", false, subQuery)
        assertEquals("A NOT IN (SELECT Z FROM Y)", filter9.buildFilter(SQLQueryBuilder()))
    }

    @Test
    fun testAndFilter() {
        val subQuery = SQLQueryBuilder().field("Z").table("Y")
        val filter: InFilter = InFilter.create("A", false, "B").and("A", true, listOf("B", "C", "D")).and("A", null.toString()).and("A", subQuery)
        assertEquals("A NOT IN (B) AND A IN (B, C, D) AND A IN (null) AND A IN (SELECT Z FROM Y)", filter.buildFilter(SQLQueryBuilder()))
    }

    @Test
    fun testOrFilter() {
        val subQuery = SQLQueryBuilder().field("Z").table("Y")
        val filter: InFilter = InFilter.create("A", "B").or("A", true, listOf("B", "C", "D")).or("A", false, null.toString()).or("C", "D").or("A", subQuery)
        assertEquals("A IN (B) OR A IN (B, C, D) OR A NOT IN (null) OR C IN (D) OR A IN (SELECT Z FROM Y)", filter.buildFilter(SQLQueryBuilder()))
    }

    @Test
    fun testComplexFilter() {
        val subQuery = SQLQueryBuilder().field("Z").table("Y")
        val filter = InFilter.create("A", "B").or("A", false, listOf("B", "C", "D")).or("A", true, subQuery).and("A", null.toString())
            .and("A", false, subQuery)
        assertEquals(
            "(A IN (B) OR A NOT IN (B, C, D) OR A IN (SELECT Z FROM Y)) AND A IN (null) AND A NOT IN (SELECT Z FROM Y)",
            filter.buildFilter(SQLQueryBuilder())
        )
    }

    @Test
    fun testBadArgument() {
        assertFailure {
            InFilter("A", true, emptyList<Any>())
            error("Should have thrown an exception")
        }
    }

//    @Test
//    fun testClone() {
//        val subQuery: JPAQueryBuilder<*> = JPAQueryBuilder<Any?>()
//        subQuery.`object`("Y", "Y")
//        subQuery.field("Y", "Z")
//        val filter: InFilter = InFilter.create("A", "B").and("A", subQuery)
//        val clone = filter.clone()
//        assertEquals(filter.buildFilter(JPAQueryBuilder<Any?>()), clone.buildFilter(JPAQueryBuilder<Any?>()))
//    }
}